Add support for capturing viewport when rendering images

astefanutti opened the pull request:
Adds a new 'mode' option to the render API so that images can be captured
based on the viewport size when set to 'viewport' as opposed to being
captured as a page based on the content size when set to 'page'. The
'mode' option defaults to 'page' for backward compatibility.

Fixes #10619

astefanutti commented:
SlimerJS has an [`onlyViewport`](https://github.com/laurentj/slimerjs/blob/76851eb550b284eb0aae1b1880cb79c751c2f12c/src/modules/webpageUtils.jsm#L312) option that does exactly that. That may be valuable to align the proposed API on this.

sodabrew commented:
This looks good - I am currently faking it with clipRect, but it's not perfect. +1 for matching up the API with SlimerJS.

astefanutti commented:
I've just updated the PR to align the API on that of SlimerJS [`onlyViewport`](https://github.com/laurentj/slimerjs/blob/76851eb550b284eb0aae1b1880cb79c751c2f12c/src/modules/webpageUtils.jsm#L312).

jeantil commented:
This would be a fix for the issue I filed https://github.com/ariya/phantomjs/issues/12875, since your issue predates mine and you have a patch, I'll close mine :)

Note that the JsonWireProtocol standard has been transfered over to W3C. After some debate, it was updated so that capturing a screenshot should always honor viewport (w3c/webdriver#98 and SeleniumHQ/selenium#646 (in the comments))

istr commented:
The changeset alleviates the root cause of the problem, but it does not fix it completely.

At the core of the problem (and related problems) is changing the viewport for full page images. 

```C++
m_customWebPage->setViewportSize(contentsSize);
```

This leads to the CSS / Layout engine propagating wrong view port dimensions. The effect is that anything that relies on view port dimensions (sticky stuff, vh / vw dimensions) will never render correctly in "full page" capture mode, regardless if set by JavaScript or CSS only.

Digging deeper into WebKit reveals that there is a function ```setPaintsEntireContents();``` in ```QWebFrameAdapter``` and in ```ScrollView``` that could be used to render the full page, but without changing the viewport. In conjunction with a WebKit based tiled backing store, ```clipsRepaints``` and ```delegatesScrolling``` this could lead to a clean solution for this problem and maybe the Qt render work-around could be removed that way, too.

However, this would need to enable access to the ```webcore``` module of webkit through qt, and even though I read ```src/qt/qtwebkit/Tools/qmake/README``` thoroughly and trying to find an example how to do this properly, I could not yet figure out how to achieve this.

Any ideas?

astefanutti commented:
@istr I agree though the issue with `setViewportSize` is _orthogonal_ to that pull request. I would suggest your create a dedicated issue so that we can iterate on it separately.

ariya commented on src/webpage.cpp:
```diff
@@ -1057,18 +1058,23 @@ QString WebPage::renderBase64(const QByteArray& format)
 }
 
-QImage WebPage::renderImage()
+QImage WebPage::renderImage(bool onlyViewport)
```
This is a no no. See http://ariya.ofilabs.com/2011/08/hall-of-api-shame-boolean-trap.html for details.

adeodhar-quark replied:
Thanks @astefanutti! This solves my problem especially with 'responsive' pages. Earlier I had trouble rendering screenshots from sites like - http://html5up.net.

But I have to agree with @ariya, the bool can mislead. Also, the name onlyViewport is not self explanatory, it doesn't tell me what this parameter is going to do.

astefanutti replied:
@adeodhar-quark thanks for the feedback. I agree with @ariya too. I'll refactor the PR ASAP, probably adding a separate method with a meaningful name.

adeodhar-quark replied:
Hi Antonin (@astenfanutti), @ariya, can you please let me know if this issue  has this been fixed in phantomjs 2.1?  Or has the release changed to 2.2 (i hope it hasnt') for this issue? 
Thanks!

astefanutti replied:
@adeodhar-quark I didn't get a chance to work on it. So still pending ATM. I'll work on it ASAP.

adeodhar-quark replied:
Thanks @astefanutti much appreciated!

zackw commented:
@adeodhar-quark 

> Or has the release changed to 2.2 (i hope it hasnt') for this issue? 

2.1 has already been released, so, yes, this will hopefully be in 2.2.

astefanutti committed:
Add support for capturing viewport when rendering images

Adds a new 'onlyViewport' option to the render API so that images can be captured
based on the viewport size when set to 'true' as opposed to being
captured as a page based on the content size when set to 'false'. The
'onlyViewport' option defaults to 'false' for backward compatibility.

Align on SlimerJS

Fixes #10619

astefanutti commented:
@ariya I've just updated the PR to use an enum instead of a boolean, as per your review comment. Do not hesitate if you have other points.

Vitallium commented:
@astefanutti could you please update your PR to reflect recent changes?

astefanutti commented:
@Vitallium I've just rebased it at HEAD revision.

Vitallium commented:
@astefanutti thank you! Will land it ASAP.

Vitallium commented:
@astefanutti and landed! Thank you!

astefanutti commented:
@Vitallium thanks for the merge and the heads-up!
